### PR TITLE
MISUV-7587: opt-out-user-request api call

### PR DESCRIPTION
**Pull request by james-king-hmrc**
MISUV-7587: opt-out-user-request api call

### CODE REVIEW

File: app/models/optOut/OptOutMessageResponse.scala [hunk 1]
```diff
@@ -18,8 +18,28 @@ package models.optOut
...
 import models.itsaStatus.StatusDetail
+import play.api.libs.json.{Format, Json}
 
 case class YearStatusDetail(taxYear: TaxYear, statusDetail: StatusDetail)
 case class OptOutMessageResponse(oneYearOptOut: Boolean = false, taxYears: Array[TaxYear] = Array()) {
   def oneYearOptOutTaxYear: TaxYear = taxYears(0)
 }
+
+
+case class OptOutUpdateReason(code: Int) {
```
Comment (avelx) [inline app/models/optOut/OptOutMessageResponse.scala:29]: good candidate to use value class instead of plain case class

Comment (james-king-hmrc) [reply app/models/optOut/OptOutMessageResponse.scala:29]: turns out I don't need it, but will do if I introduce it again

File: test/connectors/OptOutConnectorTest.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+          List(ErrorItem("INVALID_TAXABLE_ENTITY_ID",
+          "Submission has not passed validation. Invalid parameter taxableEntityId."))
+        )
+        val httpResponse = HttpResponse(BAD_REQUEST, Json.toJson(apiFailResponse), Map(CorrelationIdHeader -> Seq("123")))
+
+        setupHttpClientMock[OptOutUpdateRequest](connector.getUrl(taxableEntityId))(apiRequest, httpResponse)
+
+        val result: Future[OptOutUpdateResponse] = connector.requestOptOutForTaxYear(taxYear, taxableEntityId)
+
+        Await.result(result, 10.seconds)
```
Comment (avelx) [inline test/connectors/OptOutConnectorTest.scala:111]: Please avoid using Await.result anywhere in the code base
https://www.scalatest.org/user_guide/async_testing

File: test/mocks/MockHttp.scala [hunk 3]
```diff
@@ -33,6 +33,7 @@ import scala.concurrent.{ExecutionContext, Future}
 
 trait MockHttp extends UnitSpec with BeforeAndAfterEach {
 
+  //todo: refactor: rename mockHttpGet to mockHttp
```
Comment (ian-hmrc) [inline test/mocks/MockHttp.scala:36]: Was there a reason for not doing this?

File: app/connectors/OptOutConnector.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  def getUrl(taxableEntityId: String): String =
+    s"${appConfig.itvcProtectedService}/income-tax/itsa-status/update/$taxableEntityId"
+
+  def requestOptOutForTaxYear(taxYear: TaxYear, taxableEntityId: String)
+                             (implicit headerCarrier: HeaderCarrier): Future[OptOutUpdateResponse] = {
+
+    val body = OptOutUpdateRequest(taxYear = taxYear.toString)
+
+    http.PUT[OptOutUpdateRequest, HttpResponse](
+      getUrl(taxableEntityId), body, Seq[(String, String)]()
```
Comment (ian-hmrc) [inline app/connectors/OptOutConnector.scala:50]: I found getUrl() a little confusing in the context of a PUT... Maybe rename to just url()?

File: app/models/optOut/OptOutUpdateRequestModel.scala [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  private val defaultUpdateReason: Int = 10
+  private val successfulStatusCode: Int = NO_CONTENT
+
+  case class OptOutUpdateRequest(taxYear: String, updateReason: Int = defaultUpdateReason)
+  sealed trait OptOutUpdateResponse {
+    val statusCode: Int
+  }
+  case class OptOutUpdateResponseSuccess(correlationId: String, statusCode: Int = successfulStatusCode) extends OptOutUpdateResponse
+  case class ErrorItem(code: String, reason: String)
+  case class OptOutUpdateResponseFailure(statusCode: Int, failures: List[ErrorItem]) extends OptOutUpdateResponse
```
Comment (ian-hmrc) [inline app/models/optOut/OptOutUpdateRequestModel.scala:33]: I'm not really a fan of using polymorphic response objects like this. I think we might be more idiomatic with an Either or Future. However, I know this is a pattern in the app. Maybe one to bring up at Tech Deep Dive sometime?

File: app/connectors/OptOutConnector.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    ).map { response =>
+      response.status match {
+        case Status.NO_CONTENT =>
+          response.json.validate[OptOutUpdateResponseSuccess].fold(
```
Comment (ian-hmrc) [inline app/connectors/OptOutConnector.scala:54]: Hmm... I'm confused by this. In the case of NO_CONTENT, there shouldn't be any payload, so why are we validating it?

File: app/connectors/OptOutConnector.scala [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          valid => {
+            valid.copy(
+              statusCode = response.status,
+              correlationId = response.headers(CorrelationIdHeader).headOption.getOrElse(s"Unknown_$CorrelationIdHeader")
+            )
+          }
+        )
+        case _ =>
+          response.json.validate[OptOutUpdateResponseFailure].fold(
+            invalid => {
```
Comment (ian-hmrc) [inline app/connectors/OptOutConnector.scala:68]: According to the API spec, it looks like the error will contain `code` and `reason`, yet `OptOutUpdateResponseFailure(statusCode: Int, failures: List[ErrorItem])` does not match that. So I assume we're always going to fail validation and always use the invalid block? Is that intended?

Comment (james-king-hmrc) [reply app/connectors/OptOutConnector.scala:68]: I see that failures is an array of (code and reason) items, agree?
Screenshot 2024-05-09 at 09 36 33

File: app/models/optOut/OptOutUpdateRequestModel.scala (hunk 5 above)
Comment (james-king-hmrc) [reply app/models/optOut/OptOutUpdateRequestModel.scala:33]: I guess we still need the OptOutUpdateResponseFailure to parse the response but I can/prefer Either[ErrorCode, OptOutUpdateResponseSuccess], happy to change, shall I?

File: test/mocks/MockHttp.scala (hunk 3 above)
Comment (james-king-hmrc) [reply test/mocks/MockHttp.scala:36]: I was planning to do any refactoring needed as a separate PR at the end as it's not specifically related to this ticket, including may be the isCrystalised Boolean ticket

File: app/models/optOut/OptOutUpdateRequestModel.scala (hunk 5 above)
Comment (avelx) [reply app/models/optOut/OptOutUpdateRequestModel.scala:33]: We have to map response in the way @james-king-hmrc  implemented this, its a standard approach across FE and BE, so should be fine, Future[Either[_,_]] type is used in the up stream logic close to the core domain

File: test/mocks/MockHttp.scala (hunk 3 above)
Comment (ian-hmrc) [reply test/mocks/MockHttp.scala:36]: Cool

File: app/connectors/OptOutConnector.scala (hunk 7 above)
Comment (ian-hmrc) [reply app/connectors/OptOutConnector.scala:57]: Ah, OK. I think I mis-read that spec. Looks good with the `ErrorItem(code: String, reason: String)`.

File: app/models/optOut/OptOutUpdateRequestModel.scala (hunk 5 above)
Comment (james-king-hmrc) [reply app/models/optOut/OptOutUpdateRequestModel.scala]: yes indeed i was gonna keep OptOutUpdateResponseFailure but return Either[ErrorCode, OptOutUpdateResponseSuccess]

Comment (ian-hmrc) [reply app/models/optOut/OptOutUpdateRequestModel.scala]: Yeah. I know this is a standard pattern, but it still feels less than optimal. I'm happy to stick with it for now, but maybe this is one to discuss in a Tech Deep Dive? Feels like we should be moving away from this?

Comment (james-king-hmrc) [reply app/models/optOut/OptOutUpdateRequestModel.scala]: i think returning Either helps in exiting (when in error) the monad bind loop in a more standard way without the extra step of checking if the response was a failure or not

File: app/connectors/OptOutConnector.scala (hunk 4 above)
Comment (ian-hmrc) [reply app/connectors/OptOutConnector.scala]: OK. I can see that has been updated.

File: app/connectors/OptOutConnector.scala (hunk 6 above)
Comment (ian-hmrc) [reply app/connectors/OptOutConnector.scala]: OK. I can see that has been addressed. :)

**Review by ian-hmrc — APPROVED** [review]
LGTM

**Pull request merged**